feat(page): exposed additional info for ConsoleMessage object

tompere opened the pull request:
This patch add additional method to ConsoleMessage object sent on console event. Currently, added URL and line number to additionalInfo.

Fixes #3029

tompere committed:
feat(page): exposed additional info for ConsoleMessage object

This patch add additional method to ConsoleMessage object sent on console event. Currently, added URL and line number to additionalInfo.

Fixes #3029

JoelEinbinder commented on lib/Page.js:
```diff
@@ -179,11 +179,11 @@ class Page extends EventEmitter {
    * @param {!Protocol.Log.entryAddedPayload} event
    */
   _onLogEntryAdded(event) {
```
This is not the only entry point for console messages. There is also Runtime.consoleAPICalled. It has line number and url contained in the stack trace.

tompere replied:
can you please elaborate?
I'm looking into [`Runtime.consoleAPICalled`](https://chromedevtools.github.io/devtools-protocol/tot/Runtime#event-consoleAPICalled) and I don't see such details, nor a reference to [`ExceptionDetails`](https://chromedevtools.github.io/devtools-protocol/tot/Runtime#type-ExceptionDetails)

JoelEinbinder replied:
It is contained in the callframes of the stacktrace: https://chromedevtools.github.io/devtools-protocol/tot/Runtime#type-CallFrame

tompere replied:
done

JoelEinbinder commented on docs/api.md:
```diff
@@ -2168,6 +2169,11 @@ puppeteer.launch().then(async browser => {
 
 [ConsoleMessage] objects are dispatched by page via the ['console'](#event-console) event.
 
+#### consoleMessage.additionalInfo()
```
Why not just put the url and line number on the console message itself?

tompere replied:
you mean extending the api? same as `type` and `args`?

JoelEinbinder replied:
yes

tompere replied:
as I commented in the [origin issue](https://github.com/GoogleChrome/puppeteer/issues/3029#issuecomment-427738035):
> This is under the assumption that additionalInfo is a subset of LogEntry optional params, thus future requests for additional data can easily added

makes sense?

tompere replied:
also the fact the lineNumber and url are optionals, made me feel less comfortable in adding a dedicated and explicit API (as it will return `undefined` many times...)

JoelEinbinder replied:
Having `additionalInfo` raises some difficult questions. Which values are considered core, and which are considered additional? Can additional info be null, or will it be an empty object? Even if we have reasoning, its not obvious to a user of the API. They would have to check the documentation every time.

That being said, it might make sense to group line, column, and url under a `location` object if they are often all null or all present.

tompere replied:
done. `location` it is.

JoelEinbinder commented on lib/Page.js:
```diff
@@ -495,7 +495,12 @@ class Page extends EventEmitter {
+    let location;
+    if (event.stackTrace && event.stackTrace.callFrames) {
+      const {url, lineNumber, columnNumber} = event.stackTrace.callFrames[0];
+      location = {url, lineNumber, columnNumber};
```
nit: don't unconditionally add the properties if they might be undefined. Otherwise a user doing `"url" in location` will be misled.

tompere replied:
just a clarification about how defensive puppeteer code should be:

Can i rely on devtools protocol API for optional/mandatory return values?
e.g.
* [Runtime#type-CallFrame](https://chromedevtools.github.io/devtools-protocol/tot/Runtime#type-CallFrame) always return `{url, lineNumber, columnNumber}`.
* [Runtime#event-consoleAPICalled](https://chromedevtools.github.io/devtools-protocol/tot/Runtime#event-consoleAPICalled) might not return `StackTrace`. But if returned, it will always return `Array{CallFrame}`.

JoelEinbinder replied:
You can rely on the values that the protocol will provide, but there might be extra values alongside those values. So you should build your own object instead of serving the raw protocol object.

tompere replied:
done

JoelEinbinder commented on lib/Page.js:
```diff
@@ -517,7 +522,7 @@ class Page extends EventEmitter {
    * @param {!Array<!Puppeteer.JSHandle>} args
    */
-  _addConsoleMessage(type, args) {
+  _addConsoleMessage(type, args, location) {
```
jsdoc

tompere replied:
done

tompere committed:
feat(page): exposed additional info for ConsoleMessage object

This patch add additional method to ConsoleMessage object sent on console event. Currently, added URL and line number to additionalInfo.

Fixes #3029

aslushnikov reviewed (approved):
In it goes!

tompere commented:
👏
